File: sdk/go/webhook.go
package vettid

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// AuthEvent represents an authentication webhook event.
type AuthEvent struct {
	RequestID  string        `json:"request_id"`
	UserID     string        `json:"user_id"`
	Status     RequestStatus `json:"status"`
	Timestamp  time.Time     `json:"timestamp"`
	SessionKey string        `json:"session_key,omitempty"`
}

// AuthzEvent represents an authorization webhook event.
type AuthzEvent struct {
	RequestID string        `json:"request_id"`
	UserID    string        `json:"user_id"`
	Action    string        `json:"action"`
	Resource  string        `json:"resource"`
	Status    RequestStatus `json:"status"`
	Timestamp time.Time     `json:"timestamp"`
	ExpiresAt *time.Time    `json:"expires_at,omitempty"`
}

// ContractEvent represents a contract webhook event.
type ContractEvent struct {
	ContractID string    `json:"contract_id"`
	UserID     string    `json:"user_id"`
	ServiceID  string    `json:"service_id"`
	OfferingID string    `json:"offering_id"`
	EventType  string    `json:"contract_event"` // "accepted", "cancelled", "expired"
	Timestamp  time.Time `json:"timestamp"`
}

// CallEvent represents a call webhook event.
type CallEvent struct {
	CallID    string                 `json:"call_id"`
	RequestID string                 `json:"request_id"`
	UserID    string                 `json:"user_id"`
	Status    CallStatus             `json:"status"`
	Timestamp time.Time              `json:"timestamp"`
	Answer    map[string]interface{} `json:"answer,omitempty"`
	Duration  int                    `json:"duration_seconds,omitempty"`
}

// PaymentEvent represents a payment webhook event.
type PaymentEvent struct {
	RequestID      string        `json:"request_id"`
	UserID         string        `json:"user_id"`
	Status         PaymentStatus `json:"status"`
	Timestamp      time.Time     `json:"timestamp"`
	PaymentID      string        `json:"payment_id,omitempty"`
	TransactionRef string        `json:"transaction_id,omitempty"`
	FailureReason  string        `json:"failure_reason,omitempty"`
}

// SecretEvent represents a secret operation webhook event.
type SecretEvent struct {
	RequestID string        `json:"request_id"`
	UserID    string        `json:"user_id"`
	SecretID  string        `json:"secret_id,omitempty"`
	Operation string        `json:"operation"` // "stored", "retrieved", "deleted"
	Status    RequestStatus `json:"status"`
	Timestamp time.Time     `json:"timestamp"`
	Data      []byte        `json:"-"` // Decoded from base64
}

// Handler function types
type (
	AuthHandler     func(event *AuthEvent) error
	AuthzHandler    func(event *AuthzEvent) error
	ContractHandler func(event *ContractEvent) error
	CallHandler     func(event *CallEvent) error
	PaymentHandler  func(event *PaymentEvent) error
	SecretHandler   func(event *SecretEvent) error
)

// WebhookRouter routes webhook events to handlers.
type WebhookRouter struct {
	signingSecret    string
	authHandlers     []AuthHandler
	authzHandlers    []AuthzHandler
	contractHandlers []ContractHandler
	callHandlers     []CallHandler
	paymentHandlers  []PaymentHandler
	secretHandlers   []SecretHandler
}

// NewWebhookRouter creates a new webhook router.
func NewWebhookRouter(signingSecret string) *WebhookRouter {
	return &WebhookRouter{
		signingSecret: signingSecret,
	}
}

// OnAuth registers an authentication event handler.
func (r *WebhookRouter) OnAuth(handler AuthHandler) *WebhookRouter {
	r.authHandlers = append(r.authHandlers, handler)
	return r
}

// OnAuthz registers an authorization event handler.
func (r *WebhookRouter) OnAuthz(handler AuthzHandler) *WebhookRouter {
	r.authzHandlers = append(r.authzHandlers, handler)
	return r
}

// OnContract registers a contract event handler.
func (r *WebhookRouter) OnContract(handler ContractHandler) *WebhookRouter {
	r.contractHandlers = append(r.contractHandlers, handler)
	return r
}

// OnCall registers a call event handler.
func (r *WebhookRouter) OnCall(handler CallHandler) *WebhookRouter {
	r.callHandlers = append(r.callHandlers, handler)
	return r
}

// OnPayment registers a payment event handler.
func (r *WebhookRouter) OnPayment(handler PaymentHandler) *WebhookRouter {
	r.paymentHandlers = append(r.paymentHandlers, handler)
	return r
}

// OnSecret registers a secret operation event handler.
func (r *WebhookRouter) OnSecret(handler SecretHandler) *WebhookRouter {
	r.secretHandlers = append(r.secretHandlers, handler)
	return r
}

// VerifySignature verifies the webhook signature.
func (r *WebhookRouter) VerifySignature(payload []byte, signature string) bool {
	if r.signingSecret == "" {
		return true // No verification if secret not configured
	}

	mac := hmac.New(sha256.New, []byte(r.signingSecret))
	mac.Write(payload)
	expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	return hmac.Equal([]byte(expected), []byte(signature))
}

// Handle processes an incoming webhook.
func (r *WebhookRouter) Handle(payload []byte, signature string) error {
	// Verify signature if configured
	if r.signingSecret != "" && signature != "" {
		if !r.VerifySignature(payload, signature) {
			return fmt.Errorf("invalid webhook signature")
		}
	}

	// Parse payload
	var data map[string]interface{}
	if err := json.Unmarshal(payload, &data); err != nil {
		return fmt.Errorf("invalid JSON payload: %w", err)
	}

	eventType, ok := data["event_type"].(string)
	if !ok || eventType == "" {
		return fmt.Errorf("missing event_type in payload")
	}

	switch eventType {
	case "auth":
		return r.handleAuth(data)
	case "authz":
		return r.handleAuthz(data)
	case "contract":
		return r.handleContract(data)
	case "call":
		return r.handleCall(data)
	case "payment":
		return r.handlePayment(data)
	case "secret":
		return r.handleSecret(data)
	default:
		return fmt.Errorf("unknown event type: %s", eventType)
	}
}

func (r *WebhookRouter) handleAuth(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &AuthEvent{
		RequestID:  data["request_id"].(string),
		UserID:     data["user_id"].(string),
		Status:     RequestStatus(data["status"].(string)),
		Timestamp:  timestamp,
		SessionKey: getStringOr(data, "session_key", ""),
	}

	for _, handler := range r.authHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

func (r *WebhookRouter) handleAuthz(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &AuthzEvent{
		RequestID: data["request_id"].(string),
		UserID:    data["user_id"].(string),
		Action:    data["action"].(string),
		Resource:  data["resource"].(string),
		Status:    RequestStatus(data["status"].(string)),
		Timestamp: timestamp,
	}
	if expiresAt, ok := data["expires_at"].(string); ok && expiresAt != "" {
		t, _ := time.Parse(time.RFC3339, expiresAt)
		event.ExpiresAt = &t
	}

	for _, handler := range r.authzHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

func (r *WebhookRouter) handleContract(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &ContractEvent{
		ContractID: data["contract_id"].(string),
		UserID:     data["user_id"].(string),
		ServiceID:  data["service_id"].(string),
		OfferingID: data["offering_id"].(string),
		EventType:  data["contract_event"].(string),
		Timestamp:  timestamp,
	}

	for _, handler := range r.contractHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

func (r *WebhookRouter) handleCall(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &CallEvent{
		CallID:    data["call_id"].(string),
		RequestID: getStringOr(data, "request_id", ""),
		UserID:    data["user_id"].(string),
		Status:    CallStatus(data["status"].(string)),
		Timestamp: timestamp,
	}
	if answer, ok := data["answer"].(map[string]interface{}); ok {
		event.Answer = answer
	}
	if duration, ok := data["duration_seconds"].(float64); ok {
		event.Duration = int(duration)
	}

	for _, handler := range r.callHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

func (r *WebhookRouter) handlePayment(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &PaymentEvent{
		RequestID:      data["request_id"].(string),
		UserID:         data["user_id"].(string),
		Status:         PaymentStatus(data["status"].(string)),
		Timestamp:      timestamp,
		PaymentID:      getStringOr(data, "payment_id", ""),
		TransactionRef: getStringOr(data, "transaction_id", ""),
		FailureReason:  getStringOr(data, "failure_reason", ""),
	}

	for _, handler := range r.paymentHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

func (r *WebhookRouter) handleSecret(data map[string]interface{}) error {
	timestamp, _ := time.Parse(time.RFC3339, data["timestamp"].(string))
	event := &SecretEvent{
		RequestID: data["request_id"].(string),
		UserID:    data["user_id"].(string),
		SecretID:  getStringOr(data, "secret_id", ""),
		Operation: data["operation"].(string),
		Status:    RequestStatus(data["status"].(string)),
		Timestamp: timestamp,
	}

	// Decode base64 data if present
	if dataStr, ok := data["data"].(string); ok && dataStr != "" {
		// Remove potential "sha256=" prefix for signature verification compatibility
		if strings.HasPrefix(dataStr, "sha256=") {
			dataStr = strings.TrimPrefix(dataStr, "sha256=")
		}
		decoded, err := base64.StdEncoding.DecodeString(dataStr)
		if err == nil {
			event.Data = decoded
		}
	}

	for _, handler := range r.secretHandlers {
		if err := handler(event); err != nil {
			return err
		}
	}
	return nil
}

File: sdk/nodejs/src/webhook.ts
/**
 * Webhook handling utilities for VettID Service Vault
 */

import {
  WebhookEventType,
  WebhookPayload,
  AuthResponse,
  AuthzResponse,
  DataResponse,
  ConnectionContract,
  CallResult,
  PaymentResult,
  SecretResult,
} from './types';

/**
 * Event handler function type
 */
export type WebhookHandler<T = unknown> = (data: T) => void | Promise<void>;

/**
 * Webhook event handlers
 */
export interface WebhookHandlers {
  'auth.response'?: WebhookHandler<AuthResponse>;
  'authz.response'?: WebhookHandler<AuthzResponse>;
  'data.response'?: WebhookHandler<DataResponse>;
  'contract.created'?: WebhookHandler<ConnectionContract>;
  'contract.cancelled'?: WebhookHandler<ConnectionContract>;
  'call.accepted'?: WebhookHandler<CallAcceptedEvent>;
  'call.rejected'?: WebhookHandler<CallRejectedEvent>;
  'call.ended'?: WebhookHandler<CallEndedEvent>;
  'call.missed'?: WebhookHandler<CallMissedEvent>;
  'payment.approved'?: WebhookHandler<PaymentApprovedEvent>;
  'payment.denied'?: WebhookHandler<PaymentDeniedEvent>;
  'payment.completed'?: WebhookHandler<PaymentResult>;
  'payment.failed'?: WebhookHandler<PaymentResult>;
  'secret.response'?: WebhookHandler<SecretResult>;
}

/**
 * Call accepted webhook event
 */
export interface CallAcceptedEvent {
  requestId: string;
  callId: string;
  userId: string;
  type: 'voice' | 'video';
}

/**
 * Call rejected webhook event
 */
export interface CallRejectedEvent {
  requestId: string;
  userId: string;
  reason?: string;
}

/**
 * Call ended webhook event
 */
export interface CallEndedEvent {
  callId: string;
  userId: string;
  duration?: number;
  reason?: string;
}

/**
 * Call missed webhook event
 */
export interface CallMissedEvent {
  requestId: string;
  userId: string;
}

/**
 * Payment approved webhook event
 */
export interface PaymentApprovedEvent {
  requestId: string;
  paymentId: string;
  userId: string;
  amount: number;
  currency: string;
  method: string;
}

/**
 * Payment denied webhook event
 */
export interface PaymentDeniedEvent {
  requestId: string;
  userId: string;
  reason?: string;
}

/**
 * WebhookRouter handles incoming webhook events from VettID Service Vault.
 *
 * @example
 * ```typescript
 * const router = new WebhookRouter();
 *
 * router.on('auth.response', async (data) => {
 *   if (data.status === 'approved') {
 *     // User authenticated successfully
 *     await createSession(data.userId);
 *   }
 * });
 *
 * router.on('payment.approved', async (data) => {
 *   // Process the payment
 *   await chargeCard(data.paymentId);
 * });
 *
 * // In your HTTP handler:
 * app.post('/webhooks/vettid', async (req, res) => {
 *   await router.handle(req.body);
 *   res.status(200).send('OK');
 * });
 * ```
 */
export class WebhookRouter {
  private handlers: WebhookHandlers = {};
  private catchAllHandler?: WebhookHandler<WebhookPayload>;

  /**
   * Register a handler for a specific event type.
   */
  on<K extends keyof WebhookHandlers>(
    eventType: K,
    handler: WebhookHandlers[K]
  ): this {
    this.handlers[eventType] = handler as any;
    return this;
  }

  /**
   * Register a catch-all handler for unhandled events.
   */
  onAny(handler: WebhookHandler<WebhookPayload>): this {
    this.catchAllHandler = handler;
    return this;
  }

  /**
   * Handle an incoming webhook payload.
   *
   * @param payload - The webhook payload from VettID
   */
  async handle(payload: WebhookPayload | string): Promise<void> {
    const data: WebhookPayload = typeof payload === 'string'
      ? JSON.parse(payload)
      : payload;

    // Convert timestamp if needed
    if (typeof data.timestamp === 'string') {
      data.timestamp = new Date(data.timestamp);
    }

    const handler = this.handlers[data.eventType as keyof WebhookHandlers];
    if (handler) {
      const eventData = this.transformEventData(data.eventType as WebhookEventType, data.data);
      await (handler as WebhookHandler)(eventData);
    } else if (this.catchAllHandler) {
      await this.catchAllHandler(data);
    }
  }

  /**
   * Transform raw webhook data into typed event objects.
   */
  private transformEventData(eventType: WebhookEventType, data: any): any {
    switch (eventType) {
      case 'auth.response':
        return {
          requestId: data.request_id,
          status: data.status,
          userId: data.user_id,
          timestamp: new Date(data.timestamp),
          sessionKey: data.session_key,
        } as AuthResponse;

      case 'authz.response':
        return {
          requestId: data.request_id,
          status: data.status,
          userId: data.user_id,
          action: data.action,
          resource: data.resource,
          timestamp: new Date(data.timestamp),
          expiresAt: data.expires_at ? new Date(data.expires_at) : undefined,
        } as AuthzResponse;

      case 'call.accepted':
        return {
          requestId: data.request_id,
          callId: data.call_id,
          userId: data.user_id,
          type: data.type,
        } as CallAcceptedEvent;

      case 'call.rejected':
        return {
          requestId: data.request_id,
          userId: data.user_id,
          reason: data.reason,
        } as CallRejectedEvent;

      case 'call.ended':
        return {
          callId: data.call_id,
          userId: data.user_id,
          duration: data.duration,
          reason: data.reason,
        } as CallEndedEvent;

      case 'call.missed':
        return {
          requestId: data.request_id,
          userId: data.user_id,
        } as CallMissedEvent;

      case 'payment.approved':
        return {
          requestId: data.request_id,
          paymentId: data.payment_id,
          userId: data.user_id,
          amount: data.amount,
          currency: data.currency,
          method: data.method,
        } as PaymentApprovedEvent;

      case 'payment.denied':
        return {
          requestId: data.request_id,
          userId: data.user_id,
          reason: data.reason,
        } as PaymentDeniedEvent;

      case 'secret.response':
        return {
          requestId: data.request_id,
          status: data.status,
          secretId: data.secret_id,
          // Note: actual secret data is not included in webhooks for security
        } as SecretResult;

      default:
        return data;
    }
  }
}

/**
 * Express middleware for handling VettID webhooks.
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createWebhookMiddleware, WebhookRouter } from '@vettid/service-vault-sdk';
 *
 * const app = express();
 * const router = new WebhookRouter();
 *
 * router.on('auth.response', (data) => {
 *   console.log('Auth response:', data);
 * });
 *
 * app.post('/webhooks/vettid', express.json(), createWebhookMiddleware(router));
 * ```
 */
export function createWebhookMiddleware(router: WebhookRouter) {
  return async (req: any, res: any, next?: any) => {
    try {
      await router.handle(req.body);
      res.status(200).send('OK');
    } catch (error) {
      console.error('Webhook handling error:', error);
      res.status(500).send('Internal Server Error');
      if (next) next(error);
    }
  };
}
